framework/src/vectorpostprocessors/CSVReader.C

```diff
@@ -66,16 +70,23 @@ CSVReader::CSVReader(const InputParameters & params)
 void
 CSVReader::initialize()
 {
+  // prevent user from reading the same file multiple times
+  if (_initialized)
+    mooseError("Error in " + name() + ". CSVReader cannot execute more than once per file.");
+
+  // read file and declare the vectors
   if (_column_data.empty())
   {
     _csv_reader.read();
     for (auto & name : _csv_reader.getNames())
       _column_data[name] = &declareVector(name);
   }
-  else if (!getParam<bool>("force_preic") ||
-           (getParam<bool>("force_preic") &&
-            (_fe_problem.getCurrentExecuteOnFlag() != EXEC_INITIAL)))
-    mooseError("Error in " + name() + ". CSVReader cannot execute more than once per file.");
+
+  // mark that file has been read (allowed to twice if executing in PREIC)
```

Comment:
Refresh my memory please, why is it allowed to read twice for PREIC?

Reply:
@aeslaughter for whatever reason, when executing in PREIC, it seems to be called twice - once during PREIC (before even all the usual information that's printed at the beginning of every MOOSE execution) and once again during the normal `EXEC_INITIAL` stage. I've verified this through simple `std::cout` messages. 

I don't think it should be this way, of course, but I lack a better solution right now. With a bit of guidance, I might be able to find out the true reason why this is the case.

Reply:
This is kind-of nasty. Do you have time to put in a `print_trace()` here? I'd be curious to see what we are calling it twice. That seems like a bug.

Reply:
@permcody for sure, I want to get this right! I will see if I can find out where its coming from.

Reply:
@permcody okay I ran the `print_trace()`. The problem is that the AuxGroups are overdetermined. A user object is either in the `PRE_IC` group or it is not, but it is always in either the `PRE_AUX` group or the `POST_AUX` group, since these two groups are determined based on whether `AttribPreAux` is true or false.  `FEProblemBase::computeUserObjectsInternal()` is invoked in both the `initialSetup()` and `execute()` routines, the former checks `AttribPreIC` and the latter checks `AttribPreAux` and so a UO executed during `initialSetup()` will definitely be executed once more during `execute()`.

Perhaps the `Moose::POST_AUX` group should have its own separate warehouse attribute?

Reply:
@crswong888 - I think I'm in agreement. It seems like we shouldn't be executing thing more than once for init. The Warehouse code has changed substantially since I originally added this capability. 

@rwcarlsen - Is there a quick way we can make sure that execute_on = 'initial' will only execute a UO once. The issue here is that when an IC depends on a UO, I guess we are "adding" an attribute so it gets executed in the old place (either PRE_AUX or POST_AUX) plus the PRE_IC location so that the dependency will be properly resolved. It needs to be exclusive to one of these locations.

Reply:
@permcody I think, for now, I shall hold off on this PR until there's no need for allowing it to execute in both PRE_IC and POST_AUX, since this was already made possible in my last merge.

Reply:
the FEProblemBase::execute function calls `computeUserObjects(flag, PRE/POST_AUX)` The computeControls call has an `if (flat != INITIAL)` guard - maybe we just want to put that same guard around the computeUserObject calls?

Reply:
That's a reasonable workaround. If you have a minute to push that up as a PR - it'd be interesting to see if it changes anything. First, I think there are only a handful of tests that use the PRE_IC capability and those that do probably are OK if we execute the UO one less time.

Reply:
@rwcarlsen if any tests break, I could probably lend a hand with fixing them.

Reply:
@rwcarlsen @permcody I don't think its possible to just us the (flag != INITIAL), because most UOs aren't in the `PRE_IC` group, so those still need to run for `EXEC_INITIAL`. We don't necessarily know which group a UO object belongs to until we run the query in `computeUserObjectInternal()` and see if we get anything. 

I think the only possibilities (besides a complete overhaul of the UO dependency system, which I get the feeling that this is the eventual plan), would be to either mark that a particular UO has executed, so that we know not to run that one again for all subsequent loops through `computeUserObjects()`, or create a `POST_AUX` attribute, so that each UO may be uniquely assigned to a group.

Reply:
Yeah - all these inter-system dependency hacks are suboptimal.  Just do the simplest thing that fixes the problem and add a nice verbose to the code explaining why/how it works.